app/upgrades/v16/concentrated_pool.go

```diff
@@ -0,0 +1,154 @@
+	for _, coin := range poolLiquidity {
+		if coin.Denom == desiredDenom0 {
+			foundDenom0 = true
+		} else {
+			denom1 = coin.Denom
+		}
+	}
+
+	if !foundDenom0 {
+		return nil, NoDesiredDenomInPoolError{desiredDenom0}
+	}
+
+	// TODO: confirm pre-launch that it is the same for CL as in balancer.
+	swapFee := cfmmPool.GetSwapFee(ctx)
+
+	createPoolMsg := clmodel.NewMsgCreateConcentratedPool(poolCreatorAddress, desiredDenom0, denom1, tickSpacing, exponentAtPriceOne, swapFee)
+	concentratedPool, err := poolmanagerKeeper.CreatePoolZeroLiquidityNoFee(ctx, createPoolMsg)
+	if err != nil {
+		return nil, err
+	}
+
+	return concentratedPool, nil
+}
+
+// createCanonicalConcentratedLiuqidityPoolAndMigrationLink creates a new concentrated liquidity pool from an existing
+// CFMM pool, and migrates the gauges and distribution records from the CFMM pool to the new CL pool.
+// Additionally, it creates a migration link between the CFMM pool and the CL pool and stores it in x/gamm.
+// Returns error if fails to create concentrated liquidity pool from CFMM pool.
+// Returns error if fails to get gauges for CFMM pool.
+// Returns error if fails to get gauge for the concentrated liquidity pool.
+func createCanonicalConcentratedLiuqidityPoolAndMigrationLink(ctx sdk.Context, cfmmPoolId uint64, desiredDenom0 string, keepers *keepers.AppKeepers) error {
+	concentratedPool, err := createConcentratedPoolFromCFMM(ctx, cfmmPoolId, desiredDenom0, *keepers.AccountKeeper, *keepers.GAMMKeeper, *keepers.PoolManagerKeeper)
+	if err != nil {
+		return err
+	}
+
+	// Get CFMM gauges
+	cfmmGauges, err := getGaugesForCFMMPool(ctx, *keepers.IncentivesKeeper, *keepers.PoolIncentivesKeeper, cfmmPoolId)
+	if err != nil {
+		return err
+	}
+
+	// Get distribution epoch duration.
+	distributionEpochDuration := keepers.IncentivesKeeper.GetEpochInfo(ctx).Duration
+
+	// Get concentrated gauge correspondng to the distribution epoch duration.
+	concentratedGaugeId, err := keepers.PoolIncentivesKeeper.GetPoolGaugeId(ctx, concentratedPool.GetId(), distributionEpochDuration)
+	if err != nil {
+		return err
+	}
+
+	// Find the gauge that corresponds to the distribution epoch duration.
+	gaugeIdToRedirectTo := uint64(0)
+	foundDesiredGauge := false
+	for _, cfmmGauge := range cfmmGauges {
+		if cfmmGauge.DistributeTo.Duration == distributionEpochDuration {
+			gaugeIdToRedirectTo = cfmmGauge.Id
+			foundDesiredGauge = true
+		}
+	}
+	if !foundDesiredGauge {
+		return CouldNotFindGaugeToRedirectError{DistributionEpochDuration: distributionEpochDuration}
+	}
+
+	// Iterate through all the distr records, and redirect the old balancer gauge to the new concentrated gauge.
+	distrInfo := keepers.PoolIncentivesKeeper.GetDistrInfo(ctx)
+	for i, distrRecord := range distrInfo.Records {
+		if distrRecord.GaugeId == gaugeIdToRedirectTo {
+			distrInfo.Records[i].GaugeId = concentratedGaugeId
+		}
+	}
+
+	// Set the new distr info.
+	keepers.PoolIncentivesKeeper.SetDistrInfo(ctx, distrInfo)
```

Comment:
nit: we only really need to submit the distrInfo we are changing rather than the entire records with the changed records, but this works too.

Reply:
There are no helpers that allow setting a subset of records. Only all or nothing.

This doesn't really matter for performance since there are not that many records. Also, added a test to validate that unrelated records are unchanged